File: app/src/main/java/com/innovatube/boilerplate/injection/component/ApplicationComponent.java
package com.innovatube.boilerplate.injection.component;

import android.app.Application;
import android.content.Context;

import com.innovatube.boilerplate.data.DataManager;
import com.innovatube.boilerplate.data.remote.InnovatubeService;
import com.innovatube.boilerplate.injection.ApplicationContext;
import com.innovatube.boilerplate.injection.module.ApplicationModule;

import javax.inject.Singleton;

import dagger.Component;
import io.realm.Realm;
import retrofit2.Retrofit;

/**
 * Created by TOIDV on 4/4/2016.
 */

@Singleton
@Component(modules = ApplicationModule.class)
public interface ApplicationComponent {

    @ApplicationContext
    Context context();

    Application application();

    Retrofit retrofit();

    InnovatubeService inploiService();

    DataManager dataManager();

    Realm realm();


}

File: app/src/main/java/com/innovatube/boilerplate/util/di/components/ApplicationComponent.kt
package com.innovatube.boilerplate.util.di.components

import android.app.Application
import com.innovatube.boilerplate.util.di.modules.ActivityModule
import com.innovatube.boilerplate.util.di.modules.ApiModule
import com.innovatube.boilerplate.util.di.modules.ApplicationModule
import com.innovatube.boilerplate.util.di.modules.DatabaseModule
import com.innovatube.boilerplate.util.di.modules.DomainModule
import com.innovatube.boilerplate.util.di.modules.HeaderDataRepositoryModule
import dagger.BindsInstance
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        ApplicationModule::class,
        DomainModule::class,
        ApiModule::class,
        DatabaseModule::class,
        HeaderDataRepositoryModule::class
    ]
)
interface ApplicationComponent {
    fun plus(activityModule: ActivityModule): ActivityComponent

    @Component.Builder
    interface Builder {
        @BindsInstance
        fun application(application: Application): Builder

        fun build(): ApplicationComponent
    }
}
